Initial draft of blog post: Pulumi now targets .NET 6

Zaid-Ajaj opened the pull request.

mikhailshilkov commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,178 @@
+date: "2022-07-20"
+---
+
+In this blog post, we will talk about how Pulumi is now using [.NET 6](https://docs.microsoft.com/en-us/dotnet/core/whats-new/dotnet-6) as our default across the ecosystem. We will discuss the changes applied to templates, program structure and code generation. We also explain how Pulumi C# projects can benefit from latest features in .NET 6 and how it simplifies your programs overall. Let's dive ...
```
I would try to summarize our motivation for the change - .NET 6 is now the latest LTS version and it allows us to use a cleaner version of C#

mikhailshilkov commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,178 @@
+
+### Templates
+
+All of our .NET templates have been updated which now use `net6.0` as the target framework. This means that projects created with `pulumi new` require at least the [.NET 6 SDK](https://dotnet.microsoft.com/en-us/download/dotnet/6.0) to build and run the projects. The templates also simplified the program structure quite a bit and not longer use `MyStack` classes to define resources in case of C#  ...
```
Should we note that .NET 3.1 and 5 are still supported, but users need to change the project target manually (and potentially adjust the code too)?

Zaid-Ajaj replied:
We already mention that 3.1 is still supported in section "Backward-compatibility" so I think this is OK as is

mikhailshilkov commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,178 @@
+
+Targeting .NET 6 means that we can use features from C# 10 and here we see them in action: using [top-level statements](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/top-level-statements) that greatly simplify the entry point of the project and reduce it down to a single function call to `Deployment.RunAsync`. Inside this function is where we define our resources and return o ...
+
+The equivalent of the snippet above would have been this piece of code:
```
I would be fine if we skipped the old code. Maybe link to a github hash that has it. Otherwise viewers may be confused if they aren't reading carefully.

mikhailshilkov commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,178 @@
+
+Which is a lot more verbose and contains unnecessary ceremony. 
+
+Notice how you no longer have to define the types of your stack outsputs upfront. Instead, you simply return a dictionary of key-valued outputs. In fact, the function `Deployment.RunAsync` is very flexible. When you don't want to return outputs, simply don't:
```
```suggestion
Notice how you no longer have to define the types of your stack outputs upfront. Instead, you simply return a dictionary of key-valued outputs. In fact, the function `Deployment.RunAsync` is very flexible. When you don't want to return outputs, simply don't:
```

Frassle replied:
```suggestion
Notice how you no longer have to define the types of your stack outputs upfront. Instead, you simply return a dictionary of key-valued outputs. In fact, the function `Deployment.RunAsync` is very flexible. When you don't want to return outputs, simply don't:
```

Frassle commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,178 @@
+title: "Pulumi now targets .NET 6"
+authors: ["zaid-ajaj"]
+tags: ["dotnet", "csharp", "fsharp", "vb.net"]
+meta_desc: "Introducing a static-code analyzer for C# which provides instant feedback on common mistakes defining Pulumi resources"
```
Update meta_desc

mikhailshilkov commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,178 @@
+    // create resources
+    var bucket = new Bucket("my-bucket");
+    // perform async tasks
+    await Task.Delay(1000);
```
Show an invoke? Or something more meaningful, like an SDK call?

Frassle commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,178 @@
+
+### Templates
+
+All of our .NET templates have been updated which now use `net6.0` as the target framework. This means that projects created with `pulumi new` require at least the [.NET 6 SDK](https://dotnet.microsoft.com/en-us/download/dotnet/6.0) to build and run the projects. The templates also simplified the program structure quite a bit and not longer use `MyStack` classes to define resources in case of C#  ...
```
```suggestion
All of our .NET templates have been updated which now use `net6.0` as the target framework. This means that projects created with `pulumi new` require at least the [.NET 6 SDK](https://dotnet.microsoft.com/en-us/download/dotnet/6.0) to build and run the projects. The templates also simplified the program structure quite a bit and not longer use `MyStack` classes to define resources in case of C# and VB.NET templates. F# templates didn't change that much since these already didn't use the class approach to define infrastructure. 
```

Frassle commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
````diff
@@ -0,0 +1,178 @@
+});
+```
+
+This is how a stack looks like now with resource definitions and outputs. The [Get Started guide](https://www.pulumi.com/docs/get-started/) has been updated to follow this pattern as well.
````
```suggestion
This is how a stack with resource definitions and outputs now looks. The [Get Started guide](https://www.pulumi.com/docs/get-started/) has been updated to follow this pattern as well.
```

Frassle commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,178 @@
+
+### Backward-compatibility
+
+It is worth mentioning that although the new templates and the accompanying getting-started guides are updating to .NET 6, the main `Pulumi` SDK NuGet  package and cloud provider packages are still targeting `netcoreapp3.1`. This mean that if users are still using .NET SDK 3.1 to build and run your projects, everything will continue to work as always.
```
```suggestion
It is worth mentioning that although the new templates and the accompanying getting-started guides are updating to .NET 6, the main `Pulumi` SDK NuGet package and cloud provider packages are still targeting `netcoreapp3.1`. This mean that if users are still using .NET SDK 3.1 to build and run your projects, everything will continue to work as always.
```

mikhailshilkov commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
````diff
@@ -0,0 +1,178 @@
+        ["primaryStorageKey"] = storageAccountKeys.Apply(listStorageAccountKeysResult => listStorageAccountKeysResult.Keys[0]?.Value),
+    };
+});
+```
````
Add some sort of Conclusion and CTAs (call to action - link to getting started guide, for example).

Zaid-Ajaj commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,178 @@
+---
+title: "Pulumi now targets .NET 6"
```
Yeah that looks better! I wasn't sure how to better convey the sentiment

Zaid-Ajaj commented:
@mikhailshilkov @Frassle thank a lot for the review 🙏 I've accounted for your comments and finalized the blog post. If you could take another look, would really appreciate it 😄 

> I've updated the code snippets so that top-level `await Deployment.RunAsync` now uses `return await Deployment.RunAsync` and changed dotnet program-gen to do the same

RobbieMcKinstry commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+
+### Templates
+
+All of our .NET templates have been updated which now use `net6.0` as the target framework. This means that projects created with `pulumi new` require at least the [.NET 6 SDK](https://dotnet.microsoft.com/en-us/download/dotnet/6.0) to build and run the projects. The templates also simplified the program structure quite a bit and not longer use `MyStack` classes to define resources in case of C#  ...
```
> All of our .NET templates have been updated which now use `net6.0` as the target framework.

For what version of Pulumi? When was this released?

```suggestion
All of our .NET templates have been updated which now use `net6.0` as the target framework. This means that projects created with `pulumi new` require at least the [.NET 6 SDK](https://dotnet.microsoft.com/en-us/download/dotnet/6.0) to build and run the projects. The templates also simplified the program structure quite a bit and not longer use `MyStack` classes to define resources in case of C# and VB.NET templates. F# templates didn't change much because these already didn't use the class approach to define infrastructure.
```

Zaid-Ajaj replied:
Templates are not versioned because they are retrieved from `pulumi/templates` by any Pulumi CLI version

RobbieMcKinstry commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+
+Targeting .NET 6 means that we can use features from C# 10 and here we see them in action: using [top-level statements](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/top-level-statements) that greatly simplify the entry point of the project and reduce it down to a single function call to `Deployment.RunAsync`. Inside this function is where we define our resources and return o ...
+
+Notice how you no longer have to define the types of your stack outputs upfront. Instead, you simply return a dictionary of key-valued outputs. In fact, the function `Deployment.RunAsync` is very flexible. When you don't want to return outputs, simply don't:
```
```suggestion
Notice how you no longer have to define the types of your stack outputs upfront. Instead, you simply return a dictionary of key-valued outputs. In fact, the function `Deployment.RunAsync` is very flexible. When you don't want to return outputs, don't!
```

RobbieMcKinstry commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
````diff
@@ -0,0 +1,182 @@
+});
+```
+
+Moreover, when you want to `await` asynchronous functions alongside your resource definitions, just mark the lambda as `async`. Here is an example where we await function invokes and work with their results:
````
```suggestion
Moreover, when you want to `await` asynchronous functions alongside your resource definitions, just mark the lambda as `async`. Here is an example where we await function invokes and work with their results.
```

RobbieMcKinstry commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+
+### Code generation
+
+Pulumi provides tools to translate from several IaC languages and generates Pulumi programs in any of our supported languages. This includes:
```
```suggestion
Pulumi provides tools to translate from several IaC languages. These tools will generate Pulumi programs in any of our supported languages. This includes:
```

RobbieMcKinstry commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+
+Pulumi provides tools to translate from several IaC languages and generates Pulumi programs in any of our supported languages. This includes:
+
+- Translating [Azure Resource Manager (ARM)](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/overview) templates using[arm2pulumi](https://www.pulumi.com/arm2pulumi/)
```
```suggestion
- Translating [Azure Resource Manager (ARM)](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/overview) templates using [arm2pulumi](https://www.pulumi.com/arm2pulumi/)
```

RobbieMcKinstry commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+- Translating Kubernetes YAML using [kube2pulumi](https://www.pulumi.com/kube2pulumi/)
+- Translating [Pulumi YAML](https://www.pulumi.com/docs/intro/languages/yaml/) with the Pulumi CLI using `pulumi convert`
+
+All of these tools take their input language, convert it into the Pulumi intermediate language, then translate it into one of our supported languages. In case of C#, the code generation has been updated to generate code that uses latest features from .NET 6. It also fixed small issues that were present in the translation logic.
```
Hmm, this first clause feels a little colloquial to me, but I'm not sure how I'd modify it. Here's one suggestion.

```suggestion
These tools take their source language as input, convert it into the Pulumi intermediate language, then translate it into one of our supported languages. In the case of C#, the code generation has been updated to generate code that uses the latest features from .NET 6. It also fixed minor issues that were present in the translation logic.
```

RobbieMcKinstry commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+
+## Summary
+
+In conclusion, Microsoft has published .NET 6 as the latest Long-Term Support version. Using it with Pulumi makes for simpler and easier to write programs. You can give it a try by [installing .NET 6](https://dotnet.microsoft.com/en-us/download/dotnet/6.0) and using one of templates with `pulumi new`, then follow our updated [Get Started guide](https://www.pulumi.com/docs/get-started/) to get up  ...
```
```suggestion
.NET 6 is the latest Long-Term Support version in the Microsoft ecosystem. Our new .NET 6 support makes for simpler and easier to write Pulumi programs. You can give it a try by [installing .NET 6](https://dotnet.microsoft.com/en-us/download/dotnet/6.0) and using one of the templates with `pulumi new`, then follow our updated [Get Started guide](https://www.pulumi.com/docs/get-started/) to get up and running quickly. Our code generation tools have been upgraded to emit code that targets the latest features of C# and fixes small correctness issues in the translation logic.
```

RobbieMcKinstry commented:
Another excellent read! Great work @Zaid-Ajaj!

Zaid-Ajaj commented:
Thanks a lot @RobbieMcKinstry for the review and the suggestions. I've applied those to the blog ✅

mikhailshilkov commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+
+This is how a stack with resource definitions and outputs now looks. The [Get Started guide](https://www.pulumi.com/docs/get-started/) has been updated to follow this pattern as well.
+
+Targeting .NET 6 means that we can use features from C# 10 and here we see them in action: using [top-level statements](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/top-level-statements) that greatly simplify the entry point of the project and reduce it down to a single function call to `Deployment.RunAsync`. Inside this function is where we define our resources and return o ...
```
```suggestion
Targeting .NET 6 means that we can use features from C# 10, and here we see them in action: using [top-level statements](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/top-level-statements) that greatly simplify the entry point of the project and reduce it down to a single function call to `Deployment.RunAsync`. Inside this function, we define our resources and return outputs for the stack. Top-level statements also allow us to `await` asynchronous functions without having an explicit `Main(string[] args)` function or a `Program` class.
```

mikhailshilkov commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+
+### Backward-compatibility
+
+It is worth mentioning that although the new templates and the accompanying getting-started guides are updating to .NET 6, the main `Pulumi` SDK NuGet package and cloud provider packages are still targeting `netcoreapp3.1`. This mean that if users are still using .NET SDK v3.1 or v5.0 to build and run your projects, everything will continue to work as always.
```
```suggestion
It is worth mentioning that although the new templates and the accompanying getting-started guides are updating to .NET 6, the main `Pulumi` SDK NuGet package and cloud provider packages are still targeting `netcoreapp3.1`. This means that if users are still using .NET SDK v3.1 or v5.0 to build and run your projects, everything will continue to work as always.
```

mikhailshilkov commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+
+### Code generation
+
+Pulumi provides tools to translate from several IaC languages. These tools will generate Pulumi programs in any of our supported languages. This includes:
```
```suggestion
Pulumi provides tools to translate from several IaC languages. These tools will generate Pulumi programs in any of our supported languages:
```

mikhailshilkov commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+- Translating [Azure Resource Manager (ARM)](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/overview) templates using [arm2pulumi](https://www.pulumi.com/arm2pulumi/)
+- Translating Terrafrom code using [tf2pulumi](https://www.pulumi.com/tf2pulumi/)
+- Translating Kubernetes YAML using [kube2pulumi](https://www.pulumi.com/kube2pulumi/)
+- Translating [Pulumi YAML](https://www.pulumi.com/docs/intro/languages/yaml/) with the Pulumi CLI using `pulumi convert`
```
I would put convert at the top of the list. Partially because it's unclear when we will actually update other tools (should we mention that it's WIP?)

Zaid-Ajaj replied:
Mentioned that these tools will pick the update soon and that for now, the codegen updates are only in Pulumi CLI with `pulumi convert`

mikhailshilkov commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+Pulumi provides tools to translate from several IaC languages. These tools will generate Pulumi programs in any of our supported languages. This includes:
+
+- Translating [Azure Resource Manager (ARM)](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/overview) templates using [arm2pulumi](https://www.pulumi.com/arm2pulumi/)
+- Translating Terrafrom code using [tf2pulumi](https://www.pulumi.com/tf2pulumi/)
```
```suggestion
- Translating Terraform code using [tf2pulumi](https://www.pulumi.com/tf2pulumi/)
```

Zaid-Ajaj replied:
Nice catch!

mikhailshilkov commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+- Translating Kubernetes YAML using [kube2pulumi](https://www.pulumi.com/kube2pulumi/)
+- Translating [Pulumi YAML](https://www.pulumi.com/docs/intro/languages/yaml/) with the Pulumi CLI using `pulumi convert`
+
+These tools take their source language as input, convert it into the Pulumi intermediate language, then translate it into one of our supported languages. In the case of C#, the code generation has been updated to generate code that uses the latest features from .NET 6. It also fixed minor issues that were present in the translation logic.
```
```suggestion
These tools take their source language as input, convert it into the Pulumi intermediate language, then translate it into one of our supported languages. In the case of C#, we have updated the code generation to produce code that uses the latest features from .NET 6. We've also fixed minor issues that were present in the translation logic.
```

mikhailshilkov commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
````diff
@@ -0,0 +1,182 @@
+pulumi convert --language csharp --out csharp-from-yaml
+```
+
+This command generates a full C# project from the YAML program into the directory `./csharp-from-yaml`, then it restores the dependencies and builds the project. Now if you examine `Program.cs`, it contains the converted program and follows the succinct C# 10 syntax that targets .NET 6:
````
```suggestion
This command generates a full C# project from the YAML program into the directory `./csharp-from-yaml`, then restores the dependencies and builds the project. Now, if you examine `Program.cs`, it contains the converted program and follows the succinct C# 10 syntax that targets .NET 6:
```

mikhailshilkov commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+
+## Summary
+
+.NET 6 is the latest Long-Term Support version in the Microsoft ecosystem. Our new .NET 6 support makes for simpler and easier to write Pulumi programs. You can give it a try by [installing .NET 6](https://dotnet.microsoft.com/en-us/download/dotnet/6.0) and using one of the templates with `pulumi new`, then follow our updated [Get Started guide](https://www.pulumi.com/docs/get-started/) to get up ...
```
```suggestion
.NET 6 is the latest Long-Term Support version in the Microsoft ecosystem. Our new .NET 6 support makes it simpler and easier to write Pulumi programs. You can give it a try by [installing .NET 6](https://dotnet.microsoft.com/en-us/download/dotnet/6.0) and using one of the templates with `pulumi new`, then follow our updated [Get Started guide](https://www.pulumi.com/docs/get-started/) to get up and running quickly. We updated our code generation tools to emit code targeting the latest features of C# and fixed minor correctness issues in the translation logic.
```

mikhailshilkov reviewed (approved):
Looks great! Took another pass on grammar and stylistic suggestions - lots are subjective so feel free to ignore if you don't like them.

Zaid-Ajaj commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+---
+title: "Pulumi improved experience with .NET 6"
```
Yeah that's ever better. Also updated the thumbnail for it ✅

Zaid-Ajaj commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+title: "Pulumi improved experience with .NET 6"
+authors: ["zaid-ajaj"]
+tags: ["dotnet", "csharp", "fsharp", "vb.net"]
+meta_desc: "Talking about the improved experience with Pulumi projects running on .NET 6, how it simplifies program structure and upgrades generated C# code."
```
Applied this change, thanks!

Zaid-Ajaj commented on themes/default/content/blog/pulumi-targets-dotnet-6/index.md:
```diff
@@ -0,0 +1,182 @@
+tags: ["dotnet", "csharp", "fsharp", "vb.net"]
+meta_desc: "Talking about the improved experience with Pulumi projects running on .NET 6, how it simplifies program structure and upgrades generated C# code."
+meta_image: thumbnail.png
+date: "2022-07-21"
```
Done ✅

mikhailshilkov reviewed (approved):
🎉 LGTM
